autoPyTorch/data/tabular_feature_validator.py

```diff
@@ -39,87 +123,54 @@ def _fit(
-                    remainder="passthrough"
-                )
+            X = cast(pd.DataFrame, X)
 
-                # Mypy redefinition
-                assert self.encoder is not None
-                self.encoder.fit(X)
-
-                # The column transformer reoders the feature types - we therefore need to change
-                # it as well
-                # This means columns are shifted to the right
-                def comparator(cmp1: str, cmp2: str) -> int:
-                    if (
-                        cmp1 == 'categorical' and cmp2 == 'categorical'
-                        or cmp1 == 'numerical' and cmp2 == 'numerical'
-                    ):
-                        return 0
-                    elif cmp1 == 'categorical' and cmp2 == 'numerical':
-                        return -1
-                    elif cmp1 == 'numerical' and cmp2 == 'categorical':
-                        return 1
-                    else:
-                        raise ValueError((cmp1, cmp2))
-
-                self.feat_type = sorted(
-                    self.feat_type,
-                    key=functools.cmp_to_key(comparator)
-                )
+            self.all_nan_columns = set([column for column in X.columns if X[column].isna().all()])
 
-                self.categories = [
-                    # We fit an ordinal encoder, where all categorical
-                    # columns are shifted to the left
-                    list(range(len(cat)))
-                    for cat in self.encoder.transformers_[0][1].categories_
-                ]
+            categorical_columns, numerical_columns, feat_type = self._get_columns_info(X)
+
+            self.enc_columns = categorical_columns
+
+            preprocessors = get_tabular_preprocessors()
+            self.column_transformer = _create_column_transformer(
+                preprocessors=preprocessors,
+                numerical_columns=numerical_columns,
+                categorical_columns=categorical_columns,
+            )
+
+            # Mypy redefinition
+            assert self.column_transformer is not None
+            self.column_transformer.fit(X)
 
+            # The column transformer reorders the feature types
+            # therefore, we need to change the order of columns as well
+            # This means categorical columns are shifted to the left
+
+            self.feat_type = sorted(
+                feat_type,
+                key=functools.cmp_to_key(self._comparator)
+            )
+
+            # differently to categorical_columns and numerical_columns,
+            # this saves the index of the column.
             for i, type_ in enumerate(self.feat_type):
                 if 'numerical' in type_:
                     self.numerical_columns.append(i)
                 else:
                     self.categorical_columns.append(i)
 
         # Lastly, store the number of features
-        self.num_features = np.shape(X)[1]
+        self.num_features = len(X.columns)
```

Comment:
can we undo this change? As X can also be a sparse matrix and they don't have columns. The original code `np.shape(X)[1])` works for both dataframe and sparse matrices.

Reply:
```suggestion
        self.num_features = np.shape(X)[1]
```